validator/src/journal/publisher.rs

```diff
@@ -1,6 +1,58 @@
 use proto::batch::Batch;
 use std::collections::{HashSet, VecDeque};
 use std::slice::Iter;
+use std::sync::Arc;
+use std::sync::mpsc::{channel, Receiver, RecvError, SendError, Sender};
+
+/// This queue keeps track of the batch ids so that components on the edge
+/// can filter out duplicates early. However, there is still an opportunity for
+/// duplicates to make it into this queue, which is intentional to avoid
+/// blocking threads trying to put/get from the queue. Any duplicates
+/// introduced by this must be filtered out later.
+pub fn make_batch_queue() -> (IncomingBatchSender, IncomingBatchReceiver) {
+    let (sender, reciever) = channel();
+    let ids = Arc::new(HashSet::new());
```

Comment:
This needs to be an `Arc<Mutex<_>>`, since the the set is modifiable on multiple threads